File: leetcode/day6/cpp/so.cpp
#include "so.h"


int Solution::lengthOfLongestSubstring(std::string s) {
    if (s.size() == 0) return 0;

    int left = 0;
    int maxlength = 0;
    std::unordered_set<char> lookup;

    for (size_t i = 0; i < s.size(); i++)
    {
        /* code */
        while (lookup.find(s[i]) != lookup.end())
        {
            /* code */
            lookup.erase(s[left]);
            left++;
        }

        maxlength = std::max(maxlength, int(i) + 1 - left);
        lookup.insert(s[i]);
    }
    return maxlength;
}

File: leetcode/day6/python/so.py

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if (len(s) == 0):
            return 0
        left = 0
        maxlength = 0
        lookup = set()

        for i, char in enumerate(s):
            while (char in lookup):
                lookup.remove(s[left])
                left += 1
            maxlength = max(maxlength, i+1-left)
            lookup.add(char)
        return maxlength
